### ISSUE TITLE
[Big confusion] Why does server need optimizer

### ISSUE THREAD

**Issue by SamuelGong** [tensorflow/federated#834]
Hi, I have read *Communication-Efficient Learning of Deep Networks
from Decentralized Data* for the entire algorithm FedAvg, but still I can see no clues about what role does `server_optimizer_fn` play. Because what clients upload should be local model weights, they should be simply used to compute weighted average at the server end.

* The server-end optimizer is `SGD` with `lr = 1.0`. Is this a higher level of `SGD`? How does this SGD work? What if `lr` is set to other values? How about other optimizers like `ADMM`?
* Is there any document about this server-end optimizer?

Since I need to **ensure some FedAvg behaviors at server end** so would you please answer to this confusion? Thanks.

**Comment by zcharles8** [comment]
Hi Samuel. Great question. There are two important things happening here. First, instead of sending the client weights (as in the paper you mention), we broadcast the difference (the so-called `model delta` discussed in https://www.tensorflow.org/federated/api_docs/python/tff/learning/build_federated_averaging_process) between the weights broadcast to the client, and the client weights after local training.

In https://www.tensorflow.org/federated/api_docs/python/tff/learning/build_federated_averaging_process, we treat the average of the clients' `model delta` as an estimate of the gradient of the loss function, and apply it to the server model using `server_optimizer_fn()`. When `server_optimizer_fn` returns an `SGD` optimizer with learning rate 1.0, this recovers the FedAvg algorithm in _Communication-Efficient Learning of Deep Networks
from Decentralized Data_.

However, when `server_optimizer_fn` returns `SGD` with a different learning rate, or a different optimizer altogether, we can get much different (and even improved!) convergence behavior. We can use things like adaptive optimizers, momentum, and other optimization techniques. For more details on why we might want to do this, and for a full mathematical explanation of how `SGD` with learning rate 1.0 recovers FedAvg, see Algorithm 2 of https://arxiv.org/abs/2003.00295 and the discussion thereafter.

This is definitely a key concept, and might be worth highlighting more explicitly in our documentation. Thanks for the question, and let me know if you have any other questions.

**Comment by SamuelGong** [comment]
Hi Zachary,

Thanks for the surprisingly prompt and detailed answer! Now I have no question.
